File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        maxHeight = float("-inf")
        
        if height is None or len(height) == 0:
            return maxHeight
        
        low = 0
        high = len(height) - 1
        
        while low < high:
            
            maxHeight = max(maxHeight, min(height[low], height[high]) * (high - low))
            
            if height[low] <= height[high]:
                low += 1
            else:
                high -= 1
        
        return maxHeight

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int max = Integer.MIN_VALUE;
        
        int low = 0;
        int high = height.length - 1;
        
        while(low < high){
            max = Math.max(max, Math.min(height[low], height[high]) * (high - low));
            
            if(height[low] <= height[high]){
                low++;
            }
            else{
                high--;
            }
        }
        
        return max;
    }
}

// Time Compelexity: O(n)
// Space Complexity: O(1)